File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/fra_sed/melosys_eessi_melding/MelosysEessiMeldingMapper.java
package no.nav.melosys.eessi.service.sed.mapper.fra_sed.melosys_eessi_melding;

import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

import no.nav.melosys.eessi.kafka.producers.model.*;
import no.nav.melosys.eessi.models.sed.SED;
import no.nav.melosys.eessi.service.sed.helpers.LandkodeMapper;

public interface MelosysEessiMeldingMapper {

    default MelosysEessiMelding map(String aktoerId, SED sed, String rinaDokumentID, String rinaSaksnummer,
                                    String sedType, String bucType, String avsenderID, String landkode,
                                    String journalpostID, String dokumentID, String gsakSaksnummer,
                                    boolean sedErEndring, String sedVersjon) {
        var melosysEessiMelding = new MelosysEessiMelding();
        melosysEessiMelding.setSedId(rinaDokumentID);
        melosysEessiMelding.setRinaSaksnummer(rinaSaksnummer);
        melosysEessiMelding.setAvsender(new Avsender(avsenderID, LandkodeMapper.mapTilNavLandkode(landkode)));
        melosysEessiMelding.setJournalpostId(journalpostID);
        melosysEessiMelding.setDokumentId(dokumentID);
        melosysEessiMelding.setGsakSaksnummer(gsakSaksnummer != null ? Long.parseLong(gsakSaksnummer) : null);
        melosysEessiMelding.setAktoerId(aktoerId);
        melosysEessiMelding.setYtterligereInformasjon(sed.getNav().getYtterligereinformasjon());

        melosysEessiMelding.setSedType(sedType);
        melosysEessiMelding.setBucType(bucType);

        if (inneholderStatsborgerskap(sed)) {
            melosysEessiMelding.setStatsborgerskap(
                mapStatsborgerskap(sed.getNav().getBruker().getPerson().hentStatsborgerksapsliste())
            );
        }

        if (sed.getNav() != null && sed.getNav().getArbeidssted() != null) {
            melosysEessiMelding.setArbeidssteder(sed.getNav().getArbeidssted().stream().map(Arbeidssted::new).collect(Collectors.toList()));
        }

        if (sed.getNav() != null && sed.getNav().getArbeidsland() != null) {
            melosysEessiMelding.setArbeidsland(sed.getNav().getArbeidsland().stream().map(Arbeidsland::new).collect(Collectors.toList()));
        }

        melosysEessiMelding.setErEndring(sedErEndring);
        melosysEessiMelding.setSedVersjon(sedVersjon);
        return melosysEessiMelding;
    }

    default boolean inneholderStatsborgerskap(SED sed) {
        return sed.getNav() != null
            && sed.getNav().getBruker() != null
            && sed.getNav().getBruker().getPerson() != null
            && sed.getNav().getBruker().getPerson().getStatsborgerskap() != null;
    }


    default List<Statsborgerskap> mapStatsborgerskap(Collection<String> statsborgerskapListe) {
        return statsborgerskapListe.stream().map(Statsborgerskap::new).collect(Collectors.toList());
    }
}

File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/fra_sed/melosys_eessi_melding/MelosysEessiMeldingMapper.kt
package no.nav.melosys.eessi.service.sed.mapper.fra_sed.melosys_eessi_melding

import no.nav.melosys.eessi.kafka.producers.model.*
import no.nav.melosys.eessi.models.sed.SED
import no.nav.melosys.eessi.service.sed.LandkodeMapper

interface MelosysEessiMeldingMapper {
    fun map(eessiMeldingParams: EessiMeldingParams) = MelosysEessiMelding(
        sedId = eessiMeldingParams.rinaDokumentID,
        rinaSaksnummer = eessiMeldingParams.rinaSaksnummer,
        avsender = Avsender(eessiMeldingParams.avsenderID, LandkodeMapper.mapTilNavLandkode(eessiMeldingParams.landkode)),
        journalpostId = eessiMeldingParams.journalpostID,
        dokumentId = eessiMeldingParams.dokumentID,
        gsakSaksnummer = eessiMeldingParams.gsakSaksnummer?.toLongOrNull(),
        aktoerId = eessiMeldingParams.aktoerId,
        ytterligereInformasjon = eessiMeldingParams.sed.nav?.ytterligereinformasjon,
        sedType = eessiMeldingParams.sedType,
        bucType = eessiMeldingParams.bucType,
        erEndring = eessiMeldingParams.sedErEndring,
        sedVersjon = eessiMeldingParams.sedVersjon
    ).apply {
        if (inneholderStatsborgerskap(eessiMeldingParams.sed)) {
            statsborgerskap = mapStatsborgerskap(
                eessiMeldingParams.sed.nav?.bruker?.person?.hentStatsborgerksapsliste() ?: emptyList()
            )
        }
        eessiMeldingParams.sed.nav?.let { nav ->
            nav.arbeidssted?.let { arbeidsstedList ->
                arbeidssteder = arbeidsstedList.map(::Arbeidssted)
            }
            nav.arbeidsland?.let { arbeidslandList ->
                arbeidsland = arbeidslandList.map(::Arbeidsland)
            }
        }
    }

    fun inneholderStatsborgerskap(sed: SED?): Boolean =
        sed?.nav?.bruker?.person?.statsborgerskap != null

    fun mapStatsborgerskap(statsborgerskapListe: Collection<String>): List<Statsborgerskap> =
        statsborgerskapListe.map(::Statsborgerskap)
}
